### DISCUSSION TITLE
Question about gradient flow through `set_parameters` method

### DISCUSSION THREAD

**Discussion by jofrevalles** [torchsynth/torchsynth#457] [Q&A] [answered]
Hello torchsynth team,

I'm working on a project where I use `torchsynth` to generate audio based on parameters predicted by a neural network. My neural network predicts a set of parameters (as a dictionary of tensors) that are then passed to a synthesizer instance using the `set_parameters` method.

Here's a simplified version of my code structure:
```python
class AudioGenerationModule(nn.Module):
    def __init__(self, env):
        super(AudioGenerationModule, self).__init__()
        self.env = env

    def forward(self, action_params):
        # Assuming action_params is a dict with tensors that have gradients
        synth_output = self.env.synth.set_parameters(action_params)
        return synth_output
```

My question concerns the gradient flow through the `set_parameters` method. Specifically, if `action_params` contains tensors that have gradients (due to being the output of an earlier part of the neural network), will these gradients correctly flow through to the output of the synthesizer when generating audio?

In other words, I'm curious if the `set_parameters` method and the subsequent audio generation process in torchsynth are designed to maintain the computational graph, allowing gradients from a loss computed against the generated audio to flow back through to the original neural network parameters.

For now, I have not been able to maintain the gradients through the audio generation.

Thank you,
Jofre

**Comment by jofrevalles** [comment]
I will give a minimal working example. In particular, I am expecting something like this to work:
```python
# TEST 2
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchsynth.config import SynthConfig
from torchsynth.synth import Voice 

# Define the MLP model in PyTorch
class SimpleMLP(nn.Module):
    def __init__(self, features, output_features=78):
        super(SimpleMLP, self).__init__()
        self.dense1 = nn.Linear(in_features=features, out_features=features)
        self.relu = nn.ReLU()
        self.dense2 = nn.Linear(in_features=features, out_features=output_features)
        self.sigmoid = nn.Sigmoid()

    def forward(self, x):
        x = self.dense1(x)
        x = self.relu(x)
        x = self.dense2(x)
        x = self.sigmoid(x)  # Ensure output is in the range [0, 1]
        return x

batch_size = 32
synthconfig = SynthConfig(
    batch_size=batch_size,
    sample_rate=44100,
    buffer_size_seconds=4.0,
    reproducible=True,
    no_grad=False
    )
synth = Voice(synthconfig)
synth.train = True

# Initialize the MLP model
n_features = 300
mlp = SimpleMLP(features=n_features)
optimizer = optim.Adam(mlp.parameters(), lr=0.001)

# Function to predict parameters, set them in torchsynth, and synthesize sound
def predict_params_and_synthesize(model, control_input, synth):
    # Forward pass through the model
    predicted_params = model(control_input)

    synth.parameters(predicted_params)
    audio_output, parameters, is_train = synth.forward(0)

    return audio_output, parameters, predicted_params

# Generate some dummy control input
control_input = torch.randn((batch_size, n_features), requires_grad=True)

# Predict parameters and generate audio
audio_output, synth_params, model_params = predict_params_and_synthesize(mlp, control_input, synth)
target_audio = torch.randn_like(audio_output)

# Calculate loss (e.g., Mean Squared Error)
loss_function = torch.nn.MSELoss()
loss = loss_function(audio_output, target_audio)

# Backpropagate to compute gradients
loss.backward()
optimizer.step()

# Check gradients
for name, param in mlp.named_parameters():
    print(f"{name} gradient: {param.grad}")
```
But this has `None` in `mlp` gradients. I would expect the gradients from the output of the model to flow into the `audio_output`.


Thanks!

**Reply by jorshi** [reply]
Hi @jofrevalles -- thank you for your question and providing the code to help look into this! I know that torchsynth parameters should be able to receive gradients (they were when we built the library) but it seems like something has changed and they are getting blocked somewhere. I'm going to spend a few minutes looking this afternoon and will get back to you soon.

**Reply by jofrevalles** [reply]
Thank you, @jorshi !

**Comment by jorshi** [comment, marked as answer]
After digging into this a bit more, the main problem is in the torchsynth design and the way parameters are handled, which were designed for dataset generation as opposed to learning. torchsynth synths have internal `nn.Parameters` for each parameter. This is awkward b/c there isn't really a good way to set parameters externally and have gradients flow through the synth. I came up with a bit of a hacky workaround though.

There are two main things that need to happen 1) the internal torchsynth parameters need to be manually set using the output from your model and 2) the gradients of the parameters need to be manually copied to the predicted params and a second backward pass called to update grads for your model. 

Now atleast gradients are making it to the model. I think this is correct, but I haven't tested it in a real learning situation. Manually working with grads in PyTorch is a bit new to me.

Another big caveat is that gradients have not been tested on the torchsynth modules so there are almost surely going to be bad ones -- i noticed that nans were popping up so I added in a line to handle those. Might be worth testing individual modules and building up if possible.

Let me know how you get on!

```python
# Generate some dummy control input
control_input = torch.randn((batch_size, n_features))

# Predict parameters and generate audio
predicted_params = mlp(control_input)

with torch.no_grad():
    for i, param in enumerate(synth.parameters()):
        param.data = predicted_params[:, i]
        param.requires_grad = True

audio_output = synth.output()
target_audio = torch.rand_like(audio_output)

# Calculate loss (e.g., Mean Squared Error)
loss_function = torch.nn.MSELoss()
loss = loss_function(audio_output, target_audio)

# Custom backward pass
optimizer.zero_grad()

# Calculate gradients for the synth model
loss.backward()

# Extract the gradients from the synth model
synth_grad = torch.cat([p.grad.unsqueeze(-1) for p in synth.parameters()], dim=-1)

# I noticed that the gradients can be NaN -- replace them with 0
synth_grad = torch.nan_to_num(synth_grad, nan=0.0, posinf=1e8, neginf=-1e8)

# Manually pass the gradients to the model output and second backward pass
predicted_params.grad = None
predicted_params.backward(gradient=synth_grad)

optimizer.step()
```

**Reply by jofrevalles** [reply]
Thank you for the detailed answer, @jorshi ! I will give this a try and see the results. :)

I already noticed the fact that `torchsynth` uses `nn.Parameters` for each synth parameter, but I thought this was ok for this particular situation. I will try to test your workaround and see how far I can get. Thanks again for all your help!

**Comment by jofrevalles** [comment]
Okay, sorry for being so active here these past days.

I have been testing and I think that the proper way to address this problem is not to pass the `predicted_params` from the network to the synth directly. Instead, we should let the `optimizer` alone do this. As @jorshi said, the `nn.Parameter`s are not designed to being accessed and modified, and I think that the `.data` should cause the computational graph to break. So we should instead do something like this:
```python
optimizer = torch.optim.Adam([
    {'params': mlp.parameters()},
    {'params': synth.parameters(), 'lr': 0.01}
], lr=0.01)

# Generate some dummy control input
control_input = torch.randn((batch_size, n_features))

# Predict parameters and generate audio
predicted_params = mlp(control_input)

audio_output, parameters, is_train = synth.forward(0)
target_audio = torch.rand_like(audio_output)

# Calculate loss (e.g., Mean Squared Error)
loss_function = torch.nn.MSELoss()
loss = loss_function(audio_output, target_audio)

... (the rest of the code is the same)
```
where we pass the `synth.parameters()` to the `Adam` optimizer, as an example.

Let me know what do you think, @jorshi ! And thanks again for your patience :)

**Reply by jorshi** [reply]
interesting! does this work? my first thought would be that since predicted params are not passed to the synth at all, the gradients from the synth parameters won't be correct -- just becuase the gradients are copied back to the control_input on the backward pass they won't actually be reflective of the estimate made by the MLP. Have you tried with the simplest synth possible -- something with a single parameter like estimating the gain for a VCA on a static sine wave.

**Reply by jofrevalles** [reply]
Hey @jorshi ! It seems to work sometimes, the others the `nan`s in the gradients mess up the computation and we get `nan`s in the loss. Look at this example that worked:
```python
from torchsynth.module import (
    ADSR,
    ControlRateUpsample,
    MonophonicKeyboard,
    SquareSawVCO,
    VCA,
)

class SimpleSynth(AbstractSynth):
    def __init__(self, synthconfig: Optional[SynthConfig] = None):
        super().__init__(synthconfig=synthconfig)
        self.add_synth_modules(
            [
                ("keyboard", MonophonicKeyboard),
                ("adsr", ADSR),
                ("upsample", ControlRateUpsample),
                ("vco", SquareSawVCO),
                ("vca", VCA),
            ]
        ) 

    def output(self) -> torch.Tensor:
        midi_f0, note_on_duration = self.keyboard()
        envelope = self.adsr(note_on_duration)
        envelope = self.upsample(envelope)
        out = self.vco(midi_f0)
        out = self.vca(out, envelope)
        return out
```
```python
import matplotlib.pyplot as plt
from IPython.display import display, clear_output
import torch
import random
import numpy as np
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchsynth.config import SynthConfig

# Fix the seed so we can reproduce this
random.seed(0)
np.random.seed(0)
torch.manual_seed(0)

# Define the MLP model in PyTorch
class SimpleMLP(nn.Module):
    def __init__(self, features, output_features=11):
        super(SimpleMLP, self).__init__()
        self.dense1 = nn.Linear(in_features=features, out_features=features)
        self.relu = nn.ReLU()
        self.dense2 = nn.Linear(in_features=features, out_features=output_features)
        self.sigmoid = nn.Sigmoid()

    def forward(self, x):
        # x = self.dense1(x) # Disable the first layer since we get lots of `nan`s in the gradients...
        # x = self.relu(x)
        x = self.dense2(x)
        x = self.sigmoid(x)  # Ensure output is in the range [0, 1]
        return x

# Assuming the configuration and initialization of torchsynth's synth are done correctly
batch_size = 4
duration = 300. # in ms
sample_rate = 44100
audio_samples = int(sample_rate*duration/1000.)
print("audio_samples:", audio_samples)
synthconfig = SynthConfig(
    batch_size=batch_size,
    sample_rate=sample_rate,
    buffer_size_seconds=duration/1000.,
    reproducible=False,
    no_grad=False
    )

synth = SimpleSynth(synthconfig=synthconfig)

# synth = SineVCO(synthconfig)
synth.train = True

# Initialize the MLP model
n_features = audio_samples
mlp = SimpleMLP(features=n_features, output_features=len([p for p in synth.parameters()]))
optimizer = torch.optim.Adam([
    {'params': mlp.parameters()},
    {'params': synth.parameters(), 'lr': 0.001}
], lr=0.001)
optimizer.zero_grad()

# Generate some dummy control input
target_amplitude = torch.tensor([1.5, 1.1, 0.2, 0.05])
frequency = torch.tensor([440, 440, 440, 440]) 

time_range = torch.arange(0, audio_samples) / sample_rate
target_input = target_amplitude.unsqueeze(1) * torch.sin(2 * torch.pi * frequency.unsqueeze(1) * time_range.unsqueeze(0))

epochs = 35

losses = []
for epoch in range(epochs):
    plt.figure(figsize=(10, 5))
    running_loss = 0.0
    optimizer.zero_grad()
    predicted_params = mlp(target_input)
    
    audio_output, parameters, is_train = synth.forward()

    loss_function = torch.nn.MSELoss()
    loss = loss_function(audio_output, target_input)
    loss.backward()

    # Extract the gradients from the synth model
    synth_grad = torch.cat([p.grad.unsqueeze(-1) for p in synth.parameters()], dim=-1)
    predicted_params.backward(gradient=synth_grad)

    optimizer.step()
    avg_loss = loss.item()/batch_size
    losses.append(avg_loss)

    # Update the plot
    clear_output(wait=True)
    plt.plot(losses, label='Training Loss')
    plt.xlabel('Epoch')
    plt.ylabel('Loss')
    plt.title('Training Loss')
    plt.legend()
    display(plt.gcf())
    plt.clf()  # Clear figure to prepare for the next update
```
And the loss nicely decreases:
![loss_torchsynth_test_06-03-24](https://github.com/torchsynth/torchsynth/assets/61060572/c4f5c8b3-11af-41a4-a884-37bf2aa54aff)
For `epochs` > 35 the loss is always `nan`. We can also take a look if it learned the amplitude changes:
```python
audio, params, is_train = synth()
audio_init, params_init, is_train_init = SimpleSynth(synthconfig=synthconfig)()

display(HTML(f"<b>Input Audio</b>"))
print("Mean amplitudes:", [np.abs(target_input[i].detach().numpy()).mean() for i in range(batch_size)])
display(HTML(f"<b>Generated Audio Random</b>"))
print("Mean amplitudes:", [np.abs(audio_init[i].detach().numpy()).mean() for i in range(batch_size)])
display(HTML(f"<b>Generated Audio Trained</b>"))
print("Mean amplitudes:", [np.abs(audio[i].detach().numpy()).mean() for i in range(batch_size)])
```
![Screenshot from 2024-03-06 17-39-29](https://github.com/torchsynth/torchsynth/assets/61060572/9c078477-f79d-46f7-a01f-9a840000f2e9)

So I don't know, this does not seem conclusive at all.

My assumption is that there are some problems in some operations that lead to `nan` in the gradients. I have tested with simpler synths (such as just only one `SineVCO`) but it appears these do not maintain the gradients. Since the gradients have not been tested in `torchsynth` I suppose this would make sense.

Also, I think that this procedure that I am trying is like a hypernetwork controlling the parameters of another one. In this case, I think the best procedure would be not to define the parameters as `nn.Parameter`s and instead pass the information on the `forward` step. Nevertheless, I think that the workaround that you proposed is quite clever and may work in this situation

Again, thanks for your help!
